torch/nn/utils/parametrize.py

```diff
@@ -53,72 +55,172 @@ def cached():
+        #    # move data to/from the original tensor without changing its id (which is what the
+        #    # optimiser uses to track parameters)
+        #    # We also require the shape to be the same to avoid the user shooting themselves
+        #    # in the foot. This second requirement may be dropped in the future
+        #    if isinstance(Y, Tensor)
+        #      assert X.dtype == Y.dtype and X.shape == Y.shape
+        # Below we use original = X, new = Y
+
+        # Compute new
+        with torch.no_grad():
+            new = original
+            for module in reversed(self):  # type: ignore[call-overload]
+                if hasattr(module, "right_inverse"):
+                    new = module.right_inverse(new)
+                # else, we assume that it's the identity
+
+        if not isinstance(new, Tensor) and not isinstance(new, collections.Sequence):
+            raise ValueError("'right_inverse' should return a Tensor or a Sequence of tensors (list, tuple...). "
+                             f"Got {new.__class__.__name__}")
+
+        # If it is a sequence of one tensor, we unpack it
+        if not isinstance(new, Tensor) and len(new) == 1:
+            new = new[0]
+
+        # Set the number of original tensors
+        is_tensor = isinstance(new, Tensor)
+        self.ntensors = 1 if is_tensor else len(new)
+
+        # Register the tensor(s)
+        if self.ntensors == 1:
+            if original.dtype != new.dtype:
+                raise ValueError(
+                    "When `right_inverse` outputs one tensor, it may not change the dtype.\n"
+                    f"original.dtype: {original.dtype}\n"
+                    f"right_inverse(original).dtype: {new.dtype}"
+                )
+            if original.shape != new.shape:
+                raise ValueError(
+                    "When `right_inverse` outputs one tensor, it may not change the shape.\n"
+                    f"original.shape: {original.shape}\n"
+                    f"right_inverse(original).shape: {new.shape}"
+                )
+            # Set the original to original so that the user does not need to re-register the parameter
+            # manually in the optimiser
+            with torch.no_grad():
+                original.set_(new)  # type: ignore[call-overload]
+            _register_parameter_or_buffer(self, "original", original)
+        else:
+            requires_grad = original.requires_grad
+            for i, originali in enumerate(new):
+                if not isinstance(originali, Tensor):
+                    raise ValueError("'right_inverse' must return a Tensor or a Sequence of tensors "
+                                     "(list, tuple...). "
+                                     f"Got a sequence with an element {originali.__class__.__name__}.")
+
+                # If they require grad we expect the user to add the new parametesr to the optimizer
```

Comment:
I think that you would prefer to match exactly the two properties: requires_grad and Parameter or not based on the original.
Users can set requires_grad=False on Parameters, or create attributes that require gradients that are not Parameters.